include/wil/com.h

```diff
@@ -3325,6 +3325,79 @@ WI_NODISCARD auto make_range(IEnumXxx* enumPtr)
 #endif // WIL_HAS_CXX_17
 #endif // WIL_ENABLE_EXCEPTIONS
 
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP | WINAPI_PARTITION_SYSTEM)
+#ifdef __WIL_WIN32_HELPERS_INCLUDED
+
+/** RAII support for making cross-apartment (or cross process) COM calls with a timeout applied to them.
+ * When this is active any timed out calls will fail with an RPC error code such as RPC_E_CALL_CANCELED.
+ * This is a shared timeout that applies to all calls made on the current thread for the lifetime of
+ * the wil::rpc_timeout object.
+~~~
+{
+  auto timeout = wil::rpc_timeout(5000);
+  remote_object->BlockingCOMCall();
+  remote_object->AnotherBlockingCOMCall();
+}
+~~~
+Include wil/win32_helpers.h before wil/com.h to use this.
+*/
+class rpc_timeout
+{
+public:
+    rpc_timeout(DWORD timeoutInMilliseconds) : m_threadId(GetCurrentThreadId())
+    {
+        m_cancelEnablementResult = CoEnableCallCancellation(nullptr);
+        if (SUCCEEDED(m_cancelEnablementResult))
+        {
+            m_timer.reset(CreateThreadpoolTimer(&rpc_timeout::timer_callback, this, nullptr));
+            LOG_LAST_ERROR_IF_NULL(m_timer.get());
+            if (m_timer)
+            {
+                FILETIME ft = filetime::get_system_time();
+                filetime::add(ft, filetime::convert_msec_to_100ns(timeoutInMilliseconds));
+                SetThreadpoolTimer(m_timer.get(), &ft, 0, 0);
+            }
+        }
+    }
+
+    ~rpc_timeout()
+    {
+        if (SUCCEEDED(m_cancelEnablementResult))
+        {
+            CoDisableCallCancellation(nullptr);
+        }
+
+        m_timer.reset();
+    }
+
+    bool timed_out() const
+    {
+        return m_timedOut;
+    }
+
+private:
+    // Disable use of new as this class should only be declared on the stack, never the heap.
+    void* operator new(size_t) = delete;
+    void* operator new[](size_t) = delete;
+
+    static void __stdcall timer_callback(PTP_CALLBACK_INSTANCE /*instance*/, PVOID context, PTP_TIMER /*timer*/)
+    {
+        // The timer is waited upon during destruction so it is safe to rely on the this pointer in context.
+        rpc_timeout* self = static_cast<rpc_timeout*>(context);
+        if (SUCCEEDED(CoCancelCall(self->m_threadId, 0)))
+        {
+            self->m_timedOut = true;
+        }
+    }
+
+    wil::unique_threadpool_timer_nocancel m_timer;
+    HRESULT m_cancelEnablementResult{};

```

Comment:
There are several places for failure. A few things to consier:
1. Add an `operator bool` (or some sort of `HRESULT` getter) so that the caller can check if initialization succeeded.
2. Change this to a more generic `m_initResult` that will also change to the result of `GetLastError` when e.g. threadpool functions fail
3. Make non-throwing, failfast, and throwing versions of this type (advanced; not necessarily necessary)

Reply:
I keep going back and forth on this in my head and I haven't been able to converge on a firm opinion.  The old version that I'm porting from was "best effort" in terms of errors in the constructor.  This version continues that approach.  However, I also understand the desire to have stronger error detection if this is part of WIL.

Reply:
I'm _okay_ with it being best effort so long as it's documented as such. That said, I'm sure some consumers would want to know about failure. Perhaps something like the following would suffice?
```c++
operator bool() noexcept
{
    return static_cast<bool>(m_timer);
}
```

Reply:
I ended up going with option 1 and 3 both.  The nothrow variant can use the bool operator to determine if it is active.  There are throwing and failfast options that will do their thing if construction fails.